src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,325 @@
+    BaseFloat samp_freq_input = input_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_input = input_matrix.NumCols(),  // number of samples in recording
+          num_input_channel = input_matrix.NumRows();  // number of channels in recording
+    KALDI_VLOG(1) << "sampling frequency of input: " << samp_freq_input
+                  << " #samples: " << num_samp_input
+                  << " #channel: " << num_input_channel;
+    KALDI_ASSERT(input_channel < num_input_channel);
+
+    WaveData rir_wave;
+    {
+      Input ki(rir_file);
+      rir_wave.Read(ki.Stream());
+    }
+    const Matrix<BaseFloat> &rir_matrix = rir_wave.Data();
+    BaseFloat samp_freq_rir = rir_wave.SampFreq();
+    int32 num_samp_rir = rir_matrix.NumCols(),
+          num_rir_channel = rir_matrix.NumRows();
+    KALDI_VLOG(1) << "sampling frequency of rir: " << samp_freq_rir
+                  << " #samples: " << num_samp_rir
+                  << " #channel: " << num_rir_channel;
+    if (!multi_channel_output) {
+      KALDI_ASSERT(rir_channel < num_rir_channel);
+    }
+
+    Matrix<BaseFloat> noise_matrix;
+    if (!noise_file.empty()) {
+      WaveData noise_wave;
+      {
+        Input ki(noise_file);
+        noise_wave.Read(ki.Stream());
+      }
+      noise_matrix = noise_wave.Data();
+      BaseFloat samp_freq_noise = noise_wave.SampFreq();
+      int32 num_samp_noise = noise_matrix.NumCols(),
+            num_noise_channel = noise_matrix.NumRows();
+      KALDI_VLOG(1) << "sampling frequency of noise: " << samp_freq_noise
+                    << " #samples: " << num_samp_noise
+                    << " #channel: " << num_noise_channel;
+      if (multi_channel_output) {
+        KALDI_ASSERT(num_rir_channel == num_noise_channel);
+      } else {
+        KALDI_ASSERT(noise_channel < num_noise_channel);
+      }
+    }
+
+    int32 num_output_channel = (multi_channel_output ? num_rir_channel : 1);
+    Matrix<BaseFloat> out_matrix(num_output_channel, num_samp_input);
+
+    for (int32 output_channel = 0; output_channel < num_output_channel; output_channel++) {
+      Vector<BaseFloat> input(num_samp_input);
+      input.CopyRowFromMat(input_matrix, input_channel);
+
+      int32 this_rir_channel = (multi_channel_output ? output_channel : rir_channel);
+      Vector<BaseFloat> rir(num_samp_rir);
+      rir.CopyRowFromMat(rir_matrix, this_rir_channel);
+      rir.Scale(1 / MaxAbsolute(rir));
+
+      Vector<BaseFloat> noise(0);
+      if (!noise_file.empty()) {
+        noise.Resize(noise_matrix.NumCols());
+        int32 this_noise_channel = (multi_channel_output ? output_channel : noise_channel);
+        noise.CopyRowFromMat(noise_matrix, this_noise_channel);
+      }
+
+      DoReverberation(rir, &noise, &input, samp_freq_rir, snr_db, test_mode);
+
+      float input_max = MaxAbsolute(input);
+      for (int32 j = 0; j < input.Dim(); j++)
+        input(j) = (0.99 * input(j) / input_max) * pow(2, 15);
```

Comment:
@danpovey  OK, I stopped working on NormalizedWaveReader.

@tomkocse To summarize yesterday's discussion I think the TODOs are as follows ( @danpovey or @kkm000 please correct if necessary) : 

 1. Scale the RIRs, input wave and noise files with 1/2^15. Once the signal processing operations are done scale with 2^15 and call the wave writer.
 2. To avoid overshoot, compute the energy of the input wave file and scale the final reverberated+noised wav file to have this energy
 3. My suggestion : also have a constant scale option where you just scale the final reverberated+noised wav file with a user specified value.
 4. Move the unit-test to wav-reverberate-test.cc, to ensure this test is run by <code>make test</code>
 5. Remember to delete the python script get_reverber*.py as it is no longer used.